### PR TITLE
Fixes/macenko

### PR DESCRIPTION
**Pull request by AjeyPaiK**
Fixes {#16}

This PR makes the following changes:

-   Added `MacenkoNormalization` to the initial transforms.

-   Added relevant config files for it.

-   Modified the forward method of the `MacenkoNormalization` so that it's compatible with ahcore.

-   Introduced the `fit()` function that computes the staining vectors of WSIs with an option of dumping them to cache for quick computation during training.

| Normalization applied on whole image (No resampling)  | Normalisation applied on individual tiles ( No resampling) |
| ------------- | ------------- |
|image| image|

image

| SSIM between normalised WSI and Tile normalised WSI | PSNR between normalised WSI and Tile normalised |
| :-------------------------------------------------------------: | :---------------------------------------------------------: |
|  0.9499 | 29.568 |

image

### CODE REVIEW

File: ahcore/transforms/image_normalization.py [hunk 1]
```diff
@@ -1,23 +1,107 @@
...
+    return staining_parameters
+
+
+def _handle_stain_tensors(stain_tensor: torch.tensor, shape) -> torch.Tensor:
+    """
+    Parameters
+    ----------
+    stain_tensor
+    Returns
+    -------
```
Comment (JorenB) [inline ahcore/transforms/image_normalization.py:73]: FYI: incomplete docstr

File: ahcore/transforms/image_normalization.py [hunk 2]
```diff
@@ -1,23 +1,107 @@
...
+    """
+    batch, channels, height, width = shape
+    stain_tensor[stain_tensor > 255] = 255
```
Comment (JorenB) [inline ahcore/transforms/image_normalization.py:76]: Perhaps just use `.clamp()`? Or is this not conventional in ahcore

File: ahcore/transforms/image_normalization.py [hunk 3]
```diff
@@ -151,98 +342,114 @@ def find_he_components(self, optical_density_hat: torch.Tensor, eigvecs: torch.T
...
+            Optical density of the image.
+        Returns
+        -------
+        rgb_tensor: torch.Tensor
+            RGB image.
+        """
+        projection_to_reference_stains = -self._he_reference.to(od_tensor) @ od_tensor
+        optical_density_to_rgb = torch.exp(projection_to_reference_stains)
+        normalised_image_tensor = optical_density_to_rgb * self._transmitted_intensity
+        normalised_image_tensor[normalised_image_tensor > 255.] = 255.
```
Comment (JorenB) [inline ahcore/transforms/image_normalization.py:383]: could use `.clamp()` again

**Review by JorenB — APPROVED** [review]

File: ahcore/transforms/image_normalization.py [hunk 4]
```diff
@@ -1,23 +1,107 @@
...
+    if not path_to_folder.exists():
+        path_to_folder.mkdir(parents=True)
+
+    with h5py.File(path_to_folder / str(staining_parameters["wsi_name"] + ".h5"), "w") as hf:
+        for key, value in staining_parameters.items():
+            hf.create_dataset(key, data=value)
+
+
+@lru_cache(maxsize=None)
+def load_vector_from_h5_file(filenames):
```
[the remark this answers is not in the document]
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:43]: Corrected this. It should've been filename instead. Thank you! :)

File: ahcore/transforms/image_normalization.py [hunk 5]
```diff
@@ -1,23 +1,107 @@
...
+    path_to_stains = (
+                Path(os.environ.get("SCRATCH", "/tmp")) / "ahcore_cache" / "staining_parameters" / (filenames + ".h5"))
```
[the remark this answers is not in the document]
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:45]: Yes, the following is what the function does:
During training, my augmentation needs access to staining vectors which are pre-computed during start-up and dumped to disk. I save them as `.h5` files. Whereas, the `filename` is the name of the whole slide image having an extension of `.svs` or so. Therefore, I pass the stem of the filename and force the extension `.h5`. Is there a better way to do this?

File: ahcore/transforms/image_normalization.py (hunk 2 above)
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:76]: No, I can do that. Thank you

File: ahcore/transforms/image_normalization.py (hunk 1 above)
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:73]: I have added the docstrings to this. Thank you.

File: ahcore/transforms/image_normalization.py [hunk 6]
```diff
@@ -99,27 +197,121 @@ def __init__(
...
-    def convert_rgb_to_optical_density(self, image_tensor: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
-        image_tensor = image_tensor.permute(0, 2, 3, 1)
-        # calculate optical density
-        optical_density = -torch.log(
-            (image_tensor.reshape((-1, image_tensor.shape[-1])).float() + 1) / self._transmitted_intensity
+    def __compute_matrices(
+            self, image_tensor: torch.Tensor, staining_parameters: dict[str: torch.Tensor]
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """
+        Compute the H&E staining vectors and their concentration values for every pixel in the image tensor.
```
[the remark this answers is not in the document]
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:210]: This has been corrected. Thank you.

File: ahcore/transforms/image_normalization.py (hunk 3 above)
Comment (AjeyPaiK) [reply ahcore/transforms/image_normalization.py:383]: Done :D
